core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -425,12 +427,27 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
                 // Anyway: PANIC!! We stop this route, and after leaving the game in a valid state,
                 // we try again.
                 needToFindNewRoute = true
-                break
+                break // If you ever remove this break, remove the `assumeCanPassThrough` param below
             }
-            unit.removeFromTile()
-            unit.putInTile(tile)
+            unit.moveThroughTile(tile)
+            
+            // In case something goes wrong, cache the last tile we were able to end on
+            // We can assume we can pass through this tile, as we would have broken earlier
+            if (unit.movement.canMoveTo(tile, assumeCanPassThrough = true)) {
+                lastReachedEnterableTile = tile
+            }
+            
             if (unit.isDestroyed) break
         }
+        
+        if (!unit.isDestroyed)
+            unit.putInTile(lastReachedEnterableTile)
+
+        if (!unit.civInfo.gameInfo.gameParameters.godMode) {
+            unit.currentMovement -= distanceToTiles[lastReachedEnterableTile]!!.totalDistance
+            if (unit.currentMovement < 0.1) unit.currentMovement = 0f // silly floats which are "almost zero"
+            // const Epsilon, anyone?
```

Comment:
Float.MIN_VALUE sounds like one, but isn't. What we want is the difference between 1f and the next representable Float. Java has it as EPSILON in some namespace. For pure kotlin - bit manipulation to get it? Ah - there's `ulp`. So, `1f.ulp` (say that aloud three times while turning widdershins under the spring equinox new moon) ... is a good epsilon for numbers dealing in human-comprehensible orders of magnitude???